### PR TITLE
Use modern rust

### PR DESCRIPTION
**Pull request by lo48576**
TL;DR: tiny tiny improvements with no interface or functionality changes, but required rust version is bumped.

* Use Rust edition 2018
    + This requires rust 1.31 or later.
* Use idioms for edition 2018
    + Use `?` operator instead of `try_opt!` or `match`.
    + Apply [`cargo fix --edition-idioms`](https://doc.rust-lang.org/cargo/commands/cargo-fix.html).
* Apply formatter (rustfmt) and fix lint (clippy)
* Make the source modular
    + Iterators are now in another file and module.
    + Exported names and item paths are not changed, so this does not break compatibility.

These changes improve readability a little and make following contributions easier.
If you think Rust 1.31 (and edition 2018) is widespread enough, this would be worth doing.

### CODE REVIEW

**Comment by RazrFalcon** [comment]
Thanks, but I don't think that there are real reasons to bump a compiler version. v1.31 was released just half a year ago.

I think we should wait at least a year for this.
